### PR TITLE
Fix minor bugs

### LINKED ISSUE #13: Occupancy volume must greater than 0

### PR DESCRIPTION
**Pull request by RayMMond**
1. check current uow is null when rolling back
2. Create occupancy volume must greater than 0, resolve #13 
3. `AssetOccupancyCount`'s volume can not less than 0
4. refactor CanOccupy's check
5. Fix asset volume sufficient check bug in bulk category occupy

### CODE REVIEW

File: src/EasyAbp.BookingService.Application/EasyAbp/BookingService/AssetOccupancies/AssetOccupancyAppService.cs [hunk 1]
```diff
@@ -153,7 +156,10 @@ public virtual async Task CheckCreateAsync(CreateAssetOccupancyDto input)
             input.StartingTime,
             input.Duration)); // Todo: create a permission for occupying assets with a specified OccupierUserId.
 
-        await UnitOfWorkManager.Current.RollbackAsync();
+        if (UnitOfWorkManager.Current is not null)
+        {
+            await UnitOfWorkManager.Current.RollbackAsync();
```
Comment (gdlcf88) [inline src/EasyAbp.BookingService.Application/EasyAbp/BookingService/AssetOccupancies/AssetOccupancyAppService.cs:161]: Why there is a rollback?

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyCounts/AssetOccupancyCount.cs [hunk 2]
```diff
@@ -47,6 +47,8 @@ public void ChangeVolume(int changedVolume)
         {
             Volume += changedVolume;
         }
+
+        Volume = Math.Max(0, Volume);
```
Comment (gdlcf88) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyCounts/AssetOccupancyCount.cs:51]: Should throw if the volume becomes less than 0.

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/AssetOccupancyProviderBase.cs [hunk 3]
```diff
@@ -173,22 +172,29 @@ public virtual async Task<bool> CanOccupyByCategoryAsync(OccupyAssetByCategoryIn
...
                     foreach (var asset in assets)
                     {
                         var periods =
                             await GetCachedAssetDayPeriodsAsync(asset, category, dateGroup.Key, assetDayPeriods);
 
-                        if (!await IsVolumeSufficientAsync(model, periods))
+                        if (await IsVolumeSufficientAsync(model, periods))
                         {
-                            return false;
+                            periodOccupancyModels.AddRange(periods);
```
Comment (gdlcf88) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/AssetOccupancyProviderBase.cs:183]: What happens if it returns `false` directly?

**Review by gdlcf88 — CHANGES_REQUESTED** [review]

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/AssetOccupancyProviderBase.cs (hunk 3 above)
Comment (RayMMond) [reply src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/AssetOccupancyProviderBase.cs:183]: If there is a asset with insufficient volume and another asset with sufficient volume in same category, the whole bulk occupancy will fail, but it should success.

Comment (gdlcf88) [reply src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/AssetOccupancyProviderBase.cs:183]: OK. It's my code bug.

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/AssetOccupancyProviderBase.cs [hunk 4]
```diff
@@ -95,33 +94,34 @@ protected AssetOccupancyProviderBase(IServiceProvider serviceProvider)
...
-        return await PickAssetOrNullAsync(category, model) is not null;
+        if (await PickAssetOrNullAsync(category, model) is null)
+        {
+            throw new InsufficientAssetVolumeException();
+        }
     }
 
-    public virtual async Task<bool> CanBulkOccupyAsync(List<OccupyAssetInfoModel> models,
-        List<OccupyAssetByCategoryInfoModel> byCategoryModels)
+    public virtual async Task CanBulkOccupyAsync(IEnumerable<OccupyAssetInfoModel> models,
```
Comment (gdlcf88) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/AssetOccupancyProviderBase.cs:123]: Maybe `MustCanBulkOccupyAsync` or `CheckCanBulkOccupyAsync`?

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyCounts/AssetOccupancyCountVolumeCannotLessThanZeroException.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using Volo.Abp;
+
+namespace EasyAbp.BookingService.AssetOccupancyCounts;
+
+public class AssetOccupancyCountVolumeCannotLessThanZeroException : BusinessException
```
Comment (gdlcf88) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyCounts/AssetOccupancyCountVolumeCannotLessThanZeroException.cs:6]: Maybe `UnexpectedNegativeVolumeException` is better.

File: src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/BulkCanOccupyResult.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+﻿using EasyAbp.BookingService.AssetCategories;
+using EasyAbp.BookingService.Assets;
+using JetBrains.Annotations;
+
+namespace EasyAbp.BookingService.AssetOccupancyProviders;
+
+public class BulkCanOccupyResult : ICanOccupyResult
+{
+    public static BulkCanOccupyResult Success { get; } = new(true, default);
```
Comment (gdlcf88) [inline src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/BulkCanOccupyResult.cs:9]: `CanBulkOccupyResult`

**Review by gdlcf88 — CHANGES_REQUESTED** [review]

**Review by gdlcf88 — APPROVED** [review]

**Pull request merged**